learn/json_preprocessor.go

```diff
@@ -0,0 +1,54 @@
+package learn
+
+import (
+	"io"
+	"unicode"
+)
+
+type stripControlCharactersReader struct {
+	wrapped io.Reader
+}
+
+func newStripControlCharactersReader(wrapped io.Reader) stripControlCharactersReader {
+	return stripControlCharactersReader{wrapped: wrapped}
+}
+
+// Read up to len(p) bytes, removing any unescaped control characters found.
+// Removed characters do not count toward the total bytes read.  Returns the
+// number of bytes read.
+func (r stripControlCharactersReader) Read(p []byte) (n int, err error) {
+	pIdx := 0
+
+	// Read up to len(p) bytes, then discard any control characters.  Continue
+	// reading (and discarding control characters) until p is full or there are
+	// no more bytes to read.
+	for pIdx < len(p) {
+		remaining := len(p) - pIdx
+		buf := make([]byte, remaining)
```

Comment:
The `Read` interface uses the size of the buffer as an indicator of how much to read.  But you're right, rather than allocating a new buffer, I can use a slice over an existing buffer to manage the size.  Changed in b3361deaed335f184d152808a5dbb039472298bf.